### DISCUSSION TITLE
Help with string tokenization

### DISCUSSION THREAD

**Discussion by kpadmasola** [souffle-lang/souffle#2453] [Q&A] [answered]
Hi

I have an input relation where the fact looks like this: `"abc def ghi"` and I need to generate an output relation as follows
```
1, "abc"
2, "def"
3, "ghi"
```

I was trying to see if I can use the builtin string functors for this, but the documentation [link](https://souffle-lang.github.io/strings) seems to be broken. I checked `tests/example` directory as well but could not find what I needed.

Is this possible to do using Souffle? Can you please suggest where I should be looking? Sample code would be very welcome, as I am new to logic programming paradigm.

**Comment by quentin** [comment, marked as answer]
Hi,
The easiest way, if that's an option, would be to prepare the input data to match your requirement.

Otherwise, by default Souffle only offers `strlen` and `substr`. It is not very convenient but in association with `range` you can access individual characters of the input string and then recursively build sequences of consecutive characters. Something along this line.
